README
NAME
    Coro::ProcessPool - an asynchronous process pool

SYNOPSIS
        use Coro::ProcessPool;
        use Coro;

        my $pool = Coro::ProcessPool->new(
            max_procs => 4,
            max_reqs  => 100,
            include   => ['/path/to/my/task/classes', '/path/to/other/packages'],
        );

        my $double = sub { $_[0] * 2 };

        #-----------------------------------------------------------------------
        # Process in sequence
        #-----------------------------------------------------------------------
        my %result;
        foreach my $i (1 .. 1000) {
            $result{$i} = $pool->process($double, [$i]);
        }

        #-----------------------------------------------------------------------
        # Process as a batch
        #-----------------------------------------------------------------------
        my @results = $pool->map($double, 1 .. 1000);

        #-----------------------------------------------------------------------
        # Defer waiting for result
        #-----------------------------------------------------------------------
        my %deferred = map { $_ => $pool->defer($double, [$_]) } 1 .. 1000);
        foreach my $i (keys %deferred) {
            print "$i = " . $deferred{$i}->() . "\n";
        }

        #-----------------------------------------------------------------------
        # Use a "task class", implementing 'new' and 'run'
        #-----------------------------------------------------------------------
        my $result = $pool->process('Task::Doubler', 21);

        #-----------------------------------------------------------------------
        # Pipelines (work queues)
        #-----------------------------------------------------------------------
        my $pipe = $pool->pipeline;

        # Start producer thread to queue tasks
        my $producer = async {
            while (my $task = get_next_task()) {
                $pipe->queue('Some::TaskClass', $task);
            }

            # Let the pipeline know no more tasks are coming
            $pipe->shutdown;
        };

        # Collect the results of each task as they are received
        while (my $result = $pipe->next) {
            do_stuff_with($result);
        }

        $pool->shutdown;

DESCRIPTION
    Processes tasks using a pool of external Perl processes.

ATTRIBUTES
  max_procs
    The maximum number of processes to run within the process pool. Defaults
    to the number of CPUs on the ssytem.

  max_reqs
    The maximum number of tasks a worker process may run before being
    terminated and replaced with a fresh process. This is useful for tasks
    that might leak memory over time.

  include
    An optional array ref of directory paths to prepend to the set of
    directories the worker process will use to find Perl packages.

PRIVATE ATTRIBUTES
  procs_lock
    Semaphore used to control access to the worker processes. Starts
    incremented to the number of processes ("max_procs").

  num_procs
    Running total of processes that are currently running.

  procs
    Array holding the Coro::ProcessPool::Process objects.

  all_procs
  is_running
    Boolean which signals to the instance that the "shutdown" method has
    been called.

METHODS
  capacity
    Returns the number of free worker processes.

  shutdown
    Shuts down all processes and resets state on the process pool. After
    calling this method, the pool is effectively in a new state and may be
    used normally.

  process($f, $args)
    Processes code ref $f in a child process from the pool. If $args is
    provided, it is an array ref of arguments that will be passed to $f.
    Returns the result of calling $f->(@$args).

    Alternately, $f may be the name of a class implementing the methods
    "new" and "run", in which case the result is equivalent to calling
    $f->new(@$args)->run(). Note that the include path for worker processes
    is identical to that of the calling process.

    This call will yield until the results become available. If all
    processes are busy, this method will block until one becomes available.
    Processes are spawned as needed, up to "max_procs", from this method.
    Also note that the use of "max_reqs" can cause this method to yield
    while a new process is spawned.

  map($f, @args)
    Applies $f to each value in @args in turn and returns a list of the
    results. Although the order in which each argument is processed is not
    guaranteed, the results are guaranteed to be in the same order as @args,
    even if the result of calling $f returns a list itself (in which case,
    the results of that calcuation is flattened into the list returned by
    "map".

  defer($f, $args)
    Similar to "process" in ., but returns immediately. The return value is
    a code reference that, when called, returns the results of calling
    "$f-"(@$args)>.

        my $deferred = $pool->defer($coderef, [ $x, $y, $z ]);
        my $result   = $deferred->();

  pipeline
    Returns a Coro::ProcessPool::Pipeline object which can be used to pipe
    requests through to the process pool. Results then come out the other
    end of the pipe. It is up to the calling code to perform task account
    (for example, by passing an id in as one of the arguments to the task
    class).

        my $pipe = $pool->pipeline;

        my $producer = async {
            foreach my $args (@tasks) {
                $pipe->queue('Some::Class', $args);
            }

            $pipe->shutdown;
        };

        while (my $result = $pipe->next) {
            ...
        }

    All arguments to "pipeline()" are passed transparently to the
    constructor of Coro::ProcessPool::Pipeline. There is no limit to the
    number of pipelines which may be created for a pool.

    If the pool is shutdown while the pipeline is active, any tasks pending
    in Coro::ProcessPool::Pipeline::next will fail and cause the next call
    to "next()" to croak.

A NOTE ABOUT IMPORTS AND CLOSURES
    Code refs are serialized using Storable to pass them to the worker
    processes. Once deserialized in the pool process, these functions can no
    longer see the stack as it is in the parent process. Therefore, imports
    and variables external to the function are unavailable.

    Something like this will not work:

        use Foo;
        my $foo = Foo->new();

        my $result = $pool->process(sub {
            return $foo->bar; # $foo not found
        });

    Nor will this:

        use Foo;
        my $result = $pool->process(sub {
            my $foo = Foo->new; # Foo not found
            return $foo->bar;
        });

    The correct way to do this is to import from within the function:

        my $result = $pool->process(sub {
            require Foo;
            my $foo = Foo->new();
            return $foo->bar;
        });

    ...or to pass in external variables that are needed by the function:

        use Foo;
        my $foo = Foo->new();

        my $result = $pool->process(sub { $_[0]->bar }, [ $foo ]);

  Use versus require
    The "use" pragma is run a compile time, whereas "require" is evaluated
    at runtime. Because of this, the use of "use" in code passed directly to
    the "process" method can fail because the "use" statement has already
    been evaluated when the calling code was compiled.

    This will not work:

        $pool->process(sub {
            use Foo;
            my $foo = Foo->new();
        });

    This will work:

        $pool->process(sub {
            require Foo;
            my $foo = Foo->new();
        });

    If "use" is necessary (for example, to import a method or transform the
    calling code via import), it is recommended to move the code into its
    own module, which can then be called in the anonymous routine:

        package Bar;

        use Foo;

        sub dostuff {
            ...
        }

    Then, in your caller:

        $pool->process(sub {
            require Bar;
            Bar::dostuff();
        });

  If it's a problem...
    Use the task class method if the loading requirements are causing
    headaches:

        my $result = $pool->process('Task::Class', [@args]);

COMPATIBILITY
    "Coro::ProcessPool" will likely break on Win32 due to missing support
    for non-blocking file descriptors (Win32 can only call "select" and
    "poll" on actual network sockets). Without rewriting this as a network
    server, which would impact performance and be really annoying, it is
    likely this module will not support Win32 in the near future.

    The following modules will get you started if you wish to explore a
    synchronous process pool on Windows:

    Win32::Process
    Win32::IPC
    Win32::Pipe

AUTHOR
    Jeff Ober <sysread@fastmail.fm>

COPYRIGHT AND LICENSE
    Copyright (C) 2013-2015, Jeff Ober. All rights reserved.

    Redistribution and use in source and binary forms, with or without
    modification, are permitted provided that the following conditions are
    met:

    1. Redistributions of source code must retain the above copyright
    notice, this list of conditions and the following disclaimer.

    2. Redistributions in binary form must reproduce the above copyright
    notice, this list of conditions and the following disclaimer in the
    documentation and/or other materials provided with the distribution.

    3. Neither the name of the copyright holder nor the names of its
    contributors may be used to endorse or promote products derived from
    this software without specific prior written permission.

    THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
    IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
    TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
    PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
    HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
    SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED
    TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
    PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
    LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
    NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
    SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.



Changes
{{$NEXT}}

0.30 2017-11-13
- Include support was missing in AnyEvent::ProcessPool v.0.4. Bump version and
  reenabled support in calling code.
- Use common::sense

0.29 2017-11-13
- Now based on AnyEvent::ProcessPool, significantly reducing amount of
  complexity of code

0.28 2017-11-10
- Update to Coro 6.514 (experimental support for perls >= 5.22)
- Better support for closures using Data::Dump::Streamer
- Refactor, simplify, and modernize
- Dropped dependencies on Moo, Type::Tiny, and a few others
- Backward incompatible changes:
  - Coro::ProcessPool->defer returns an AnyEvent::CondVar instead of wrapping
    on in an anonymous sub which calls $cv->recv. This is to provide the caller
    with more flexibility with the result.
  - Coro::ProcessPool->(defer|process|map) no longer accept an array ref of
    arguments to be passed to the function in the worker process. Instead, the
    entire argument list is used (e.g. $pool->process(\&func, 1, 2, 3) instead
    of $pool->process(\&func, [1, 2, 3])). This is to simplify the code and
    require fewer function-local assignments, which reduces stack overhead for
    large arguments. Note that this has implications for the map method, which
    now behaves more like perl's map.
  - Coro::ProcessPool->shutdown no longer leaves the pool in a usable state;
    start a new pool instead. This change also impacts any pipelines using the
    pool (previously, the attempting to queue a task to a pipeline configured
    for a pool which had been shutdown would croak).

0.26 2015-06-09
- More reliable process management
- Cleaner shutdown for processes and pool itself
- Pipeline detects parent pool shutdown and shuts itself down
- Additional include paths may now be manually specified to the pool
- Correct handling of MSWin32 in Makefile.PL (thanks Alexandr Ciornii)
- Killed processes replaced on check-in

0.25 2015-04-17
- Fix year in previous entry to Changes
- Pipeline->next now recroaks errors
- MSWin32 now fails to build in Makefile.PL

0.24 2015-03-31
- Fixed missing argument

0.23 2015-03-31
- Set min version of Types::Standard to avoid errors on older cpan testers boxes
- Default values now use subs for older Moo versions
- Fixed bug where message was incorrectly encoded when task result was the value '0' (zero)
- Added Coro::ProcessPool::Pipeline class and pipeline method to Coro::ProcessPool

0.22 2015-03-13
- Disabled test that was running out of memory on cpan testers - sorry!

0.21 2015-03-12
- Switched from Storable to Sereal
- Added tests for oversized data

0.20 2015-03-11
- Fixed bug causing processes to hang when shutting down
- Better integration tests
- Removed queue() method; it was redundant

0.19 2015-03-10
- More refactoring of workflow
- Classes now use Moo
- Deprecated Coro::ProcessPool::Mailbox
- Better handling of child processes
- ProcessPool now starts processes immediately

0.18 2015-03-04
- Fixed ugly bug where pool did not automatically shut down

0.17 2015-03-04
- Rewrite of workflow for safe worker shutdown
- Added queue() method

0.16 2014-12-24
- Added AnyEvent as a dependency to catch obscure error on Linux/ARM systems

0.15 2014-11-14
- Processing errors now include better information
- Tasks may now be submitted with the name of a class implementing new(@args) and run()

0.14 2014-11-06
- Better support for Perl 5.8.9
- Removed dependency on Sys::Info (unsupported on too many platforms and perl versions)

0.13 2014-11-06
- Boosted Storable minimum version to 2.05 (when CODE refs were added)
- Added test to verify CODE refs usable with encode/decode routines
- Removed use of fields pragma

0.12 2014-10-31
- Messages now stringified using Coro::Storable

0.11 2014-10-15
- Fixed recursive call to readable when waiting on inbox handle readability
- map() now defined in terms of defer()

0.10 2014-08-15
- Updated documentation to clarify use of use vs require pragma
- Fixed unchecked safe_cancel in mailbox cleanup

0.09 2014-01-24
- Fixed bug when Sys::Info fails on some platforms

0.08 2014-01-16
- Fixed bad test

0.07 2014-01-10
- Added timeout parameter to process()

0.06 2013-12-18
- Fixed unintentional MSWin32 errors due to imports in tests
- More efficient I/O workflow to allow higher throughput
- Manages worker pool more efficiently

0.05 2013-12-03
- Updated documentation

0.04 2013-12-03
- Fixed documentation errors
- Added map()
- Added defer()
- +added awesomeness using Guard module

0.03 2013-11-25
- Replaced Sys::CPU with Sys::Info

0.02 2013-11-25
- Miscellaneous bugfixes
- POD updates
- Added die() statements to prevent unsafe running on MSWin32

0.01 2013-11-25
- Initial release


lib/Coro/ProcessPool.pm
package Coro::ProcessPool;
# ABSTRACT: An asynchronous pool of perl processes

use common::sense;
use Coro;
use Coro::Countdown;
use AnyEvent::ProcessPool;

sub new {
  my ($class, %param) = @_;

  my $pool = AnyEvent::ProcessPool->new(
    workers => $param{max_procs},
    limit   => $param{max_reqs},
    include => $param{include},
  );

  my $self = bless {
    pool => $pool,
    max_procs => $pool->{workers},
  }, $class;

  return $self;
}

sub join {
  my $self = shift;
  $self->{pool}->join;
}

sub defer {
  my $self = shift;
  $self->{pool}->async(@_)
}

sub process {
  my $self = shift;
  $self->{pool}->async(@_)->recv;
}

sub map {
  my ($self, $f, @args) = @_;

  # Inverse semaphore to track pending requests
  my $rem = new Coro::Countdown;

  # Queue each argument and store as an ordered list to preserve original
  # ordering of the argments
  my @cvs = map {
    $rem->up;
    $self->defer($f, $_);
  } @args;

  # Collect results, retaining original ordering by respecting the orignial
  # list index
  my @res;
  foreach my $i (0 .. $#args) {
    async_pool {
      $res[$i] = $_[0]->recv;
      $rem->down;
    } $cvs[$i];
  }

  # Wait for all requests to complete and return the result
  $rem->join;
  return @res;
}

sub pipeline {
  my $self = shift;
  return Coro::ProcessPool::Pipeline->new(pool => $self, @_);
}

=head1 SYNOPSIS

  use Coro::ProcessPool;
  use Coro;

  my $pool = Coro::ProcessPool->new(
    max_procs => 4,
    max_reqs  => 100,
    include   => ['/path/to/my/task/classes', '/path/to/other/packages'],
  );

  my $double = sub { $_[0] * 2 };

  #-----------------------------------------------------------------------
  # Process in sequence, waiting for each result in turn
  #-----------------------------------------------------------------------
  my %result;
  foreach my $i (1 .. 1000) {
    $result{$i} = $pool->process($double, $i);
  }

  #-----------------------------------------------------------------------
  # Process as a batch
  #-----------------------------------------------------------------------
  my @results = $pool->map($double, 1 .. 1000);

  #-----------------------------------------------------------------------
  # Defer waiting for result
  #-----------------------------------------------------------------------
  my %deferred;

  $deferred{$_} = $pool->defer($double, $_)
    foreach 1 .. 1000;

  # Later
  foreach my $i (keys %deferred) {
    print "$i = " . $deferred{$i}->() . "\n";
  }

  #-----------------------------------------------------------------------
  # Use a "task class" implementing 'new' and 'run'
  #-----------------------------------------------------------------------
  my $result = $pool->process('Task::Doubler', 21);

  #-----------------------------------------------------------------------
  # Pipelines (work queues)
  #-----------------------------------------------------------------------
  my $pipe = $pool->pipeline;

  # Start producer thread to queue tasks
  my $producer = async {
    while (my $task = get_next_task()) {
      $pipe->queue('Some::TaskClass', $task);
    }

    # Let the pipeline know no more tasks are coming
    $pipe->shutdown;
  };

  # Collect the results of each task as they are received
  while (my $result = $pipe->next) {
    do_stuff_with($result);
  }

  $pool->shutdown;

=head1 DESCRIPTION

Processes tasks using a pool of external Perl processes.

=head1 CONSTRUCTOR

  my $pool = Coro::ProcessPool->new(
    max_procs => 4,
    max_reqs  => 100,
    include   => ['path/to/my/packages', 'some/more/packages'],
  );

=head2 max_procs

The maximum number of processes to run within the process pool. Defaults
to the number of CPUs on the ssytem.

=head2 max_reqs

The maximum number of tasks a worker process may run before being terminated
and replaced with a fresh process. This is useful for tasks that might leak
memory over time.

=head2 include

An optional array ref of directory paths to prepend to the set of directories
the worker process will use to find Perl packages.

=head1 METHODS

=head2 join

Cedes control to the event loop until the pool has completed all remaining
tasks and woken up any threads watching them.

=head2 defer

Queues a task to be processed by the pool. Tasks may specified in either of two
forms, as a code ref or the fully qualified name of a perl class which
implements two methods, C<new> and C<run>. Any remaining arguments to C<defer>
are passed unchanged to the code ref or the C<new> method of the task class.

C<defer> will immediately return an L<AnyEvent/condvar> that will wait for and
return the result of the task (or croak if the task generated an error).

  # Using a code ref
  my $cv = $pool->defer(\&func, $arg1, $arg2, $arg3);
  my $result = $cv->recv;

  # With a task class
  my $cv = $pool->defer('Some::Task::Class', $arg1, $arg2, $arg3);
  my $result = $cv->recv;

=head2 process

Calls defer and immediately calls C<recv> on the returned condvar, returning
the result. This is useful if your workflow includes multiple threads which
share the same pool. All arguments are passed unchanged to C<defer>.

=head2 map

Like perl's C<map>, applies a code ref to a list of arguments. This method will
cede until all results have been returned by the pool, returning the result as
a list. The order of arguments and results is preserved as expected.

  my @results = $pool->map(\&func, $arg1, $arg2, $arg3);

=head2 pipeline

Returns a L<Coro::ProcessPool::Pipeline> object which can be used to pipe
requests through to the process pool. Results then come out the other end of
the pipe, not necessarily in the order in which they were queued. It is up to
the calling code to perform task accounting (for example, by passing an id in
as one of the arguments to the task class).

  my $pipe = $pool->pipeline;

  my $producer = async {
    foreach my $args (@tasks) {
      $pipe->queue('Some::Class', $args);
    }

    $pipe->shutdown;
  };

  while (my $result = $pipe->next) {
    ...
  }

All arguments to C<pipeline()> are passed transparently to the constructor of
L<Coro::ProcessPool::Pipeline>. There is no limit to the number of pipelines
which may be created for a pool.

=head1 A NOTE ABOUT IMPORTS AND CLOSURES

Code refs are serialized using L<Data::Dump::Streamer>, allowing closed over
variables to be available to the code being called in the sub-process. Mutated
variables are I<not> updated when the result is returned.

See L<Data::Dump::Streamer/Caveats-Dumping-Closures-(CODE-Refs)> for important
notes regarding closures.

=head2 Use versus require

The C<use> pragma is run at compile time, whereas C<require> is evaluated at
runtime. Because of this, the use of C<use> in code passed directly to the
C<process> method can fail in the worker process because the C<use> statement
has already been evaluated in the parent process when the calling code was
compiled.

This will not work:

  $pool->process(sub {
    use Foo;
    my $foo = Foo->new();
  });

This will work:

  $pool->process(sub {
    require Foo;
    my $foo = Foo->new();
  });

If C<use> is necessary (for example, to import a method or transform the
calling code via import), it is recommended to move the code into its own
module (or to expliticly call require and import in the subroutine), which can
then be called in the anonymous routine:

  package Bar;

  use Foo;

  sub dostuff {
    ...
  }

Then, in your caller:

  $pool->process(sub {
    require Bar;
    Bar::dostuff();
  });

Alternately, a task class may be used if dependency management is causing a
headaches:

  my $result = $pool->process('Task::Class', @args);

=head1 COMPATIBILITY

C<Coro::ProcessPool> will likely break on Win32 due to missing support for
non-blocking file descriptors (Win32 can only call C<select> and C<poll> on
actual network sockets). Without rewriting this as a network server, which
would impact performance and be really annoying, it is likely this module will
not support Win32 in the near future.

The following modules will get you started if you wish to explore a synchronous
process pool on Windows:

=over

=item L<Win32::Process>

=item L<Win32::IPC>

=item L<Win32::Pipe>

=back

=head1 SEE ALSO

=over

=item L<Coro>

=item L<AnyEvent/condvar>

=back

=cut

1;
